cpp/src/arrow/flight/client.cc

```diff
@@ -91,6 +85,134 @@ struct ClientRpc {
   }
 };
 
+/// Helper that manages Finish() of a gRPC stream.
+///
+/// When we encounter an error (e.g. could not decode an IPC message),
+/// we want to provide both the client-side error context and any
+/// available server-side context. This helper helps wrap up that
+/// logic.
+///
+/// This class protects the stream with a flag (so that Finish is
+/// idempotent), and drains the read side (so that Finish won't hang).
+template <typename Stream, typename ReadT>
```

Comment:
`ReadT` is always FlightData, right?

Reply:
Through `FinishableWritableStream` it gets templated with both `internal::FlightData` and `pb::PutResult`, so we do need the template parameter unfortunately.